Libraries/LibURL/Pattern/Component.cpp

```diff
@@ -0,0 +1,256 @@
+        // 10. Assert: part’s prefix is not the empty string or part’s suffix is not the empty string.
+        VERIFY(!part.prefix.is_empty() || !part.suffix.is_empty());
+
+        // 11. Append "(?:" to the end of result.
+        result.append("(?:"sv);
+
+        // 12. Append the result of running escape a regexp string given part’s prefix to the end of result.
+        result.append(escape_a_regexp_string(part.prefix));
+
+        // 13. Append "((?:" to the end of result.
+        result.append("((?:"sv);
+
+        // 14. Append regexp value to the end of result.
+        result.append(regexp_value);
+
+        // 15. Append ")(?:" to the end of result.
+        result.append(")(?:"sv);
+
+        // 16. Append the result of running escape a regexp string given part’s suffix to the end of result.
+        result.append(escape_a_regexp_string(part.suffix));
+
+        // 17. Append the result of running escape a regexp string given part’s prefix to the end of result.
+        result.append(escape_a_regexp_string(part.prefix));
+
+        // 18. Append "(?:" to the end of result.
+        result.append("(?:"sv);
+
+        // 19. Append regexp value to the end of result.
+        result.append(regexp_value);
+
+        // 20. Append "))*)" to the end of result.
+        result.append("))*)"sv);
+
+        // 21. Append the result of running escape a regexp string given part’s suffix to the end of result.
+        result.append(escape_a_regexp_string(part.suffix));
+
+        // 22. Append ")" to the end of result.
+        result.append(')');
+
+        // 23. If part’s modifier is "zero-or-more" then append "?" to the end of result.
+        if (part.modifier == Part::Modifier::ZeroOrMore)
+            result.append('?');
+    }
+
+    // 4. Append "$" to the end of result.
+    result.append('$');
+
+    // 5. Return (result, name list).
+    return { result.to_string_without_validation(), name_list };
+}
+
+// https://urlpattern.spec.whatwg.org/#compile-a-component
+PatternErrorOr<Component> Component::compile(Utf8View const& input, PatternParser::EncodingCallback encoding_callback, Options const& options)
+{
+    // 1. Let part list be the result of running parse a pattern string given input, options, and encoding callback.
+    auto part_list = TRY(PatternParser::parse(input, options, move(encoding_callback)));
+
+    // 2. Let (regular expression string, name list) be the result of running generate a regular expression and name
+    //    list given part list and options.
+    auto [regular_expression_string, name_list] = generate_a_regular_expression_and_name_list(part_list, options);
+
+    // 3. Let flags be an empty string.
+    auto flags = regex::RegexOptions<ECMAScriptFlags> {
+        (regex::ECMAScriptFlags)regex::AllFlags::SingleMatch
+        | (regex::ECMAScriptFlags)regex::AllFlags::Global
+        | (regex::ECMAScriptFlags)regex::AllFlags::SkipTrimEmptyMatches
+        | regex::ECMAScriptFlags::BrowserExtended
+    };
```

Comment:
:thinking:  maybe add a FIXME to de-dup this with LibJS (or at least a comment saying that's where this comes from, I was confused)